### DISCUSSION TITLE
Access idm via oidc

### DISCUSSION THREAD

**Discussion by dvv** [kanidm/kanidm#2538] [Q&A] [answered]
Hi!

I would like to try the REST API to manage group scopes by the admin persons.
I created
```
class: account
class: memberof
class: oauth2_resource_server
class: oauth2_resource_server_basic
class: object
displayname: Passport
es256_private_key_der: private_binary
memberof: idm_all_accounts@example.com
name: passport
oauth2_rs_basic_secret: hidden
oauth2_rs_origin: https://passport.example.com/
oauth2_rs_scope_map: idm_all_persons@example.com: {"email", "groups", "openid", "profile"}
oauth2_rs_token_key: hidden
spn: passport@example.com
uuid: d01a21b6-2b32-4e89-8f54-6eb53f207048
```
and setup oauth2-proxy that pipes https://passport.example.com to internal https://kanidm:8443.

I tried to sign in the passport app and it drove me to unusual "invalid webauthn authentication".
When I sign in with the same user to other applications protected in the same manner (a whoami docker image, e.g.) I succeed.

I wonder whether I go the right way to obtain a bearer token to the REST API or it should be completely different?

TIA

**Comment by yaleman** [comment]
The origin field in the server config needs to match the URL you're exposing, OAuth2 has nothing to do with the kanidm frontend.

```
#   The origin for webauthn. This is the url to the server,
#   with the port included if it is non-standard (any port
#   except 443). This must match or be a descendent of the
#   domain name you configure above. If these two items are
#   not consistent, the server WILL refuse to start!
origin = "https://idm.example.com:8443"
```

**Reply by dvv** [reply]
The origin is `https://idm.example.com` and exposed as `https://idm.example.com -> https://kanidm:8443` where `kanidm` is the name of docker container running kanidm, and it works fine.

Then I add an example OAuth2 app whoami (`kanidm system oauth2 create whoami 'WhoAmI' https://whoami.example.com`) and expose it as `https://whoami.example.com --oauth2-proxy--> http://whoami-oauth2:4180 --upstream--> http://whoami`, and it works as expected.

Next step I take by induction is to treat `kanidmd` per se as an OAuth2 app (`kanidm system oauth2 create passport 'Passport' https://passport.example.com`) and expose it as `https://passport.example.com --oauth2-proxy--> http://passport-oauth2:4180 --upstream--> https://kanidm:8443`, and signing in at https://passport.example.com fails.

**Reply by Firstyear** [reply]
What is "passport"? And why is it pointing at Kanidm?

**Reply by yaleman** [reply]
> and signing in at https://passport.example.com/ fails.

if passport.example.com is Kanidm, that's why you're getting the webauthn error. You can't expose Kanidm on multiple hostnames.

> treat kanidmd per se as an OAuth2 app

It's not an OAuth2 app.

**Reply by dvv** [reply]
It's a custom app I'm planning to create to operate on `https://idm.example.com/v1/...`.

Oh. Just noticed I do get the oauth2-proxy cookie and after signin I get redirected to `https://passport.example.com`. But since effectively it's an alias to `https://idm.example.com` with clean localstorage it redirects me to `https://passport.example.com/ui/login` and when I log in there it shows "invalid webauthn authentication", and this is indeed the right behavior since it has not come from `https://idm.example.com`.

Bump. And I can not use token obtained at `https://passport.example.com` to operate on REST API -- I believe the issuer mismatches.

So I state this is no go. Sorry for the buzz )

**Reply by Firstyear** [reply]
When you say "REST API" here, do you mean the Kanidm rest api? Is your intent to have the access_token from passport able to access the kanidm api in the background?

**Reply by dvv** [reply]
Yes, exactly.

**Reply by Firstyear** [reply]
Today that's not possible - an oidc access_token is not considered "equivalent" to a kanidm issue user token. That's why we don't allow it to work.

Generally if you need information about users there are oauth2 introspection mechansims etc. However if your intent is to have your "passport" application actual able to make changes to kanidm on behalf of users then that's a different case. 

What's the use case you have in mind here?

**Reply by dvv** [reply]
The latter.
We've got a bunch of different-aged internal apps that all manage their user base (users/groups/permissions) in different ways.
To replace that zoo of auth solutions with the unified approach we'll be using sup-scopes and custom claims, which they like to manage via UI.

**Reply by Firstyear** [reply, marked as answer]
I can see 4 possible options

1. Extend the Kanidm web UI to support the features you want, so that they can directly manage applications in kani rather than passport
2. Issue a service-account to passport that is part of the group `idm_oauth2_admins`. If a user has a valid oauth2 session, and the correct scopes in passport, they can edit the groups by proxy via the service account etc. 
3. As above, but one service-account per oauth2 using entry-managed-by access rights
4. We add support for oauth2 tokens from some clients to act as valid tokens in place of the user-auth-token. 

I will say that 4 is probably the "best" long term solution, where we can flag specific oauth2 clients to have their tokens in the same token domain as kanidm, but that is a *huge* block of work which requires https://github.com/kanidm/kanidm/issues/2513 as a start, and then a large refactoring of how we manage and issue tokens a whole. I don't see us getting to this for a while, even though I think it's a good idea.

Right now the "path of least resistance" is 1, because that benefits your users, and all other kani users. Otherwise, 2/3 solve your issue. 

Does that give you some ideas on how to proceed?

**Reply by dvv** [reply]
I've been considering 2(3) as a fallback prior to starting the discussion. Just wanted to try until I'd see a full stop.
As for 1 I do want to feed back to the project but my expertise in Rust is about `(-1..0]` and is hardly to grow in the near future. Plus it needs a consensus to exist on features etc.
So I believe I'll go 2 in waiting for 4 (no press here) feeding to 1 if something valuable is ever grown.

Thank you a lot!

**Reply by Firstyear** [reply]
No problem - sorry we didn't have the feature you wanted. Can you open an issue for 1.? I'll open the issue for 4 (since I have more ideas in mind about that).

**Reply by dvv** [reply]
I spawned #2543 -- please feel free to reword.
Also please reword this discussion title so that it were more useful )

**Reply by dvv** [reply]
Hm, as for 1 I wonder if it's a way to overload some `/ui/...` paths with own pages that implement custom logic still using privileged user-auth-token?

**Reply by Firstyear** [reply]
There is an admin ui hidden behind a feature flag :) it's just very much "not ready yet".

**Reply by dvv** [reply]
Right, I'm observing it. So if I route `https://idm.example.com/ui/admin/oauth2` to serve my own page I can re-use localstorage bearer token in api calls. Even if I'm naive it's definitely deserves a try ))
